### PR TITLE
Add the FA v2 "varlen" API

### PR DESCRIPTION
**Pull request by vgokhale**
Currently only supports the v2 API for fwd pass / inference. Causal masking, bias and dropout are not yet supported.

### CODE REVIEW

File: python/perf-kernels/flash-attention.py [hunk 1]
```diff
@@ -209,27 +289,54 @@ def _attn_fwd(
...
     BLOCK_M: tl.constexpr,
     BLOCK_DMODEL: tl.constexpr,
     BLOCK_N: tl.constexpr,
     PRE_LOAD_V: tl.constexpr,
-    NEED_PADDING: tl.constexpr,
     BIAS_TYPE: tl.constexpr,
     ENABLE_DROPOUT: tl.constexpr,
     RETURN_ENCODED_SOFTMAX: tl.constexpr
 ):
+    MQA: tl.constexpr = HQ != HK
```
Comment (xinyazhang) [inline python/perf-kernels/flash-attention.py:309]: Will become `is_mqa` when `HQ` and `HK` become `headdim_q` and `headdim_k`.

File: python/perf-kernels/flash-attention.py [hunk 2]
```diff
@@ -209,27 +289,54 @@ def _attn_fwd(
...
     start_m = tl.program_id(0)
-    off_hz = tl.program_id(1)
-    q_offset = off_hz * stride_qh
+    off_h_q = tl.program_id(1)
+    off_z = tl.program_id(2)
+    if VARLEN:
+        cu_seqlens_q_start = tl.load(cu_seqlens_q_tensor + off_z)
```
Comment (xinyazhang) [inline python/perf-kernels/flash-attention.py:314]: Can we load a size `[2]` Tensor with single `tl.load`?

Optional, can be done later since it's less relevant.

File: python/perf-kernels/flash-attention.py [hunk 3]
```diff
@@ -259,12 +366,12 @@ def _attn_fwd(
     # initialize offsets
     offs_m = start_m * BLOCK_M + tl.arange(0, BLOCK_M)
     offs_n = tl.arange(0, BLOCK_N)
-    if bias is not None:
+    if BIAS_TYPE is not None:
```
Comment (xinyazhang) [inline python/perf-kernels/flash-attention.py:369]: Ideally `BIAS_TYPE` should be a integer rather than a string.
I haven't figured out how to compile a triton kernel with string type `tl.constexpr`.

Can be done later.

File: python/perf-kernels/flash-attention.py [hunk 4]
```diff
@@ -200,7 +281,6 @@ def _attn_fwd_inner(
             encoded_softmax_block_ptr = tl.advance(encoded_softmax_block_ptr, (0, BLOCK_N))
     return acc, l_i, m_i
 
-
 @triton.jit
 def _attn_fwd(
```
Comment (xinyazhang) [inline python/perf-kernels/flash-attention.py:285]: I would remove the leading `_` to indicate this kernel will be called outside this python file.

Can be done later.

File: python/perf-kernels/flash-attention.py [hunk 5]
```diff
@@ -824,12 +924,12 @@ def backward(ctx, do, _):
...
                           (4, 16, 8192, 64),
                           (4, 16, 8080, 64),
-                          (1, 48, 16384, 64),
+                          (1, 16, 16384, 64),
                           (4, 48, 127, 64),
                           ])
 @pytest.mark.parametrize('causal', [False])
 @pytest.mark.parametrize('use_bias', [True, False])
-@pytest.mark.parametrize('bias_type', ["vector", "matrix"])
+@pytest.mark.parametrize('bias_type', ["vector"])
```
Comment (xinyazhang) [inline python/perf-kernels/flash-attention.py:932]: Just a reminder "matrix" `bias_type` is not tested here.
(I'll add a `# FIXME` or `# TODO` Here)

File: python/perf-kernels/flash-attention.py [hunk 6]
```diff
@@ -851,23 +956,116 @@ def test_op_fwd(Z, H, N_CTX, D_HEAD, causal, use_bias, bias_type, qseqlen_not_eq
...
+    q = torch.randn((total_q, HQ, D_HEAD), dtype=dtype, device="cuda").normal_(mean=0., std=0.5).requires_grad_()
+    k = torch.randn((total_k, HK, D_HEAD), dtype=dtype, device="cuda").normal_(mean=0., std=0.5).requires_grad_()
+    v = torch.randn((total_k, HK, D_HEAD), dtype=dtype, device="cuda").normal_(mean=0., std=0.5).requires_grad_()
+    sm_scale = D_HEAD**-0.5
+    input_metadata = MetaData(sm_scale=sm_scale)
+    input_metadata.set_varlen_params(cu_seqlens_q, cu_seqlens_k)
+    return q, k, v, input_metadata
+
+
+@pytest.mark.parametrize('Z, H, N_CTX, D_HEAD',
```
Comment (xinyazhang) [inline python/perf-kernels/flash-attention.py:1005]: Let's leave it as it is now but I'm expecting such patterns on UTs for full coverage,
```Python
# Copied from PyTorch
@parametrize("batch_size", [1, 8])
@parametrize("seq_len_q", [4, 8, 64, 143, 256, 512, 1024, 2048])
@parametrize("seq_len_k", [4, 8, 64, 128, 256, 587, 1024, 2048])
@parametrize("head_dim", [8, 16, 21, 32, 64, 72, 96, 128, 160, 192, 203, 256])
@parametrize("is_causal", [True, False])
@parametrize("dropout_p", [0.0, 0.22, 0.48])
@parametrize("dtype", [torch.float16, torch.bfloat16])
@parametrize("scale", [None, "l1"])
```

and for new UTs we can do it right for the first pass.

**Review by xinyazhang — APPROVED** [review]
No major problems we have to fix at this round.
All the comments can be done after the merging and splitting into multiple files with dedicated purposes.

**Review by groenenboomj — APPROVED** [review]
Aside from Xinya's comments I think this looks good.  It runs and gets what we need.  My only feedback would be to address dropout and causal masking next steps in a new PR.  Also we are fine right now but we should make sure to not add metadata as a vector for sharing context between the forwards and backwards passes as that will not work with the PyTorch integration layer without a lot of pain.

LGTM, passes it's test, so it should be good to go.

File: python/perf-kernels/flash-attention.py [hunk 7]
```diff
@@ -209,27 +289,54 @@ def _attn_fwd(
     stride_vz, stride_vh, stride_vk, stride_vn,
     stride_oz, stride_oh, stride_om, stride_on,
     stride_bz, stride_bh, stride_bm, stride_bn,
-    extra_tokens_n,
-    H,
-    seqlen_q,
-    seqlen_k,
+    cu_seqlens_q_tensor, cu_seqlens_k_tensor,
```
[the remark this answers is not in the document]
Comment (vgokhale) [reply python/perf-kernels/flash-attention.py]: Fixed.

File: python/perf-kernels/flash-attention.py [hunk 8]
```diff
@@ -33,6 +33,87 @@
...
+        self.bias_type = bias_type
+    
+    def need_causal(self):
+        self.causal = True
+
+    def need_dropout(dropout_p, return_encoded_softmax):
+        self.dropout_p = dropout_p
+        self.return_encoded_softmax = return_encoded_softmax
+
+def check_args(q, k, v, o, metadata):
```
[the remark this answers is not in the document]
Comment (vgokhale) [reply python/perf-kernels/flash-attention.py]: Fixed.

File: python/perf-kernels/flash-attention.py [hunk 9]
```diff
@@ -209,27 +289,54 @@ def _attn_fwd(
...
-    seqlen_q,
-    seqlen_k,
+    cu_seqlens_q_tensor, cu_seqlens_k_tensor,
     dropout_p,
     philox_seed,
     philox_offset_base,
     encoded_softmax,
+    max_seqlens_q, max_seqlens_k,
+    VARLEN: tl.constexpr,
+    B: tl.constexpr, HQ: tl.constexpr, HK: tl.constexpr,
```
[the remark this answers is not in the document]
Comment (vgokhale) [reply python/perf-kernels/flash-attention.py]: Fixed both.

File: python/perf-kernels/flash-attention.py (hunk 1 above)
Comment (vgokhale) [reply python/perf-kernels/flash-attention.py]: Fixed (modulo hq becoming headdim_q - it is number of q heads, not the head dim.)

File: python/perf-kernels/flash-attention.py (hunk 3 above)
Comment (vgokhale) [reply python/perf-kernels/flash-attention.py]: Fixed.

File: python/perf-kernels/flash-attention.py (hunk 4 above)
Comment (vgokhale) [reply python/perf-kernels/flash-attention.py]: Fixed.

File: python/perf-kernels/flash-attention.py (hunk 5 above)
Comment (vgokhale) [reply python/perf-kernels/flash-attention.py]: I removed this to reduce number of configs being tested and forgot to add it back. Fixed.

File: python/perf-kernels/flash-attention.py (hunk 6 above)
Comment (vgokhale) [reply python/perf-kernels/flash-attention.py:1005]: Sounds good. We can make the tests more comprehensive once we split.
